### PR TITLE
Fix race condition

### LINKED ISSUE #4792: Race condition during specialization causes multiple node processes to start
If specialization happens before webhost level java.exe is fully initialized and starting java.exe fails, 
Following code, assumes _workerRunime is same as the runtime for the language worker channel:
https://github.com/Azure/azure-functions-host/blob/af1c7ef437ac62b7b715ed7d46a1667f4f8268c6/src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs#L108-L128

Causing node process to restart instead of the java

cc @mhoeger

### LINKED ISSUE #4796: Race condition during specialization causes multiple java processes to start
If specialization happens before webhost level java.exe is fully initialized , FunctionDipatcher starts another java.exe at the  jobhost level as the handler for Worker Initialization does not invoked before specialization

### PR DESCRIPTION
**Pull request by pragnagopa**
Track webhost channels as soon as the process starts. Wait for process initialization before sending functionload requests.
Fixes #4792 Fixes #4796

### CODE REVIEW

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 1]
```diff
@@ -55,8 +55,9 @@ private async Task<ILanguageWorkerChannel> InitializeLanguageWorkerChannel(strin
             try
             {
                 languageWorkerChannel = _languageWorkerChannelFactory.CreateLanguageWorkerChannel(scriptRootPath, runtime, null, 0);
-                await languageWorkerChannel.StartWorkerProcessAsync();
                 AddOrUpdateWorkerChannels(runtime, languageWorkerChannel);
```
Comment (mhoeger) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Can we just pass in the channel ID and change this to be called something like "AddWorkerChannel"?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 2]
```diff
@@ -55,8 +55,9 @@ private async Task<ILanguageWorkerChannel> InitializeLanguageWorkerChannel(strin
...
+                await languageWorkerChannel.StartWorkerProcessAsync();
+                SetWorkerChannelState(runtime, languageWorkerChannel);
```
Comment (mhoeger) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: Maybe something like "SetInitializedWorkerChannel"?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 3]
```diff
@@ -139,15 +144,16 @@ internal void ScheduleShutdownStandbyChannels()
             }
         }
 
-        public void ShutdownChannels()
+        public async void ShutdownChannels()
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: >async void [](start = 15, length = 10)

did you mean to do async Task?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs [hunk 4]
```diff
@@ -96,26 +97,29 @@ public async Task SpecializeAsync()
...
+                        languageWorkerChannels.Remove(workerId);
+                        return true;
+                    }
                 }
             }
             return false;
         }
 
-        internal void ScheduleShutdownStandbyChannels()
+        internal async void ScheduleShutdownStandbyChannels()
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: >async void [](start = 17, length = 10)

async Task?

File: src/WebJobs.Script/Rpc/FunctionRegistration/IFunctionDispatcher.cs [hunk 5]
```diff
@@ -15,7 +15,7 @@ public interface IFunctionDispatcher : IDisposable
         // Tests if the function metadata is supported by a known language worker
         bool IsSupported(FunctionMetadata metadata, string language);
 
-        void Invoke(ScriptInvocationContext invocationContext);
+        Task Invoke(ScriptInvocationContext invocationContext);
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/FunctionRegistration/IFunctionDispatcher.cs]: >Invoke [](start = 13, length = 6)

InvokeAsync?

File: src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs [hunk 6]
```diff
@@ -229,9 +230,14 @@ public void Invoke(ScriptInvocationContext invocationContext)
             }
         }
 
-        internal IEnumerable<ILanguageWorkerChannel> GetInitializedWorkerChannels()
+        internal async Task<IEnumerable<ILanguageWorkerChannel>> GetInitializedWorkerChannels()
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/FunctionRegistration/FunctionDispatcher.cs]: >GetInitializedWorkerChannels [](start = 65, length = 28)

GetInitializatedWorkerChannelAsync?

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 1 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: This is still add or update as this method is for workerChannels maintained by the webhost channel manager

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 4 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: This is an action and ok to be async void. Fixed rest of them!

File: src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs (hunk 2 above)
Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/WebHostLanguageWorkerChannelManager.cs]: done!

**Comment by pragnagopa** [comment]
@mhoeger / @yojagad  - Addressed CR, Updated tests. Verifying on private stamp now. Thanks!

File: src/WebJobs.Script/Rpc/RpcInitializationService.cs [hunk 7]
```diff
@@ -68,7 +68,7 @@ public async Task StartAsync(CancellationToken cancellationToken)
         public Task StopAsync(CancellationToken cancellationToken)
         {
             _logger.LogDebug("Shuttingdown Rpc Channels Manager");
-            _webHostlanguageWorkerChannelManager.ShutdownChannels();
+            _webHostlanguageWorkerChannelManager.ShutdownChannelsAsync();
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/RpcInitializationService.cs]: >_webHostlanguageWorkerChannelManager.ShutdownChannelsAsync(); [](start = 12, length = 61)

await? and StopAsync could be an async Task I think

Comment (pragnagopa) [reply src/WebJobs.Script/Rpc/RpcInitializationService.cs]: Good catch! Thanks!

File: src/WebJobs.Script/Rpc/IWebHostLanguageWorkerChannelManager.cs [hunk 8]
```diff
@@ -10,12 +10,12 @@ public interface IWebHostLanguageWorkerChannelManager
     {
         Task<ILanguageWorkerChannel> InitializeChannelAsync(string language);
 
-        IEnumerable<ILanguageWorkerChannel> GetChannels(string language);
+        Dictionary<string, TaskCompletionSource<ILanguageWorkerChannel>> GetChannels(string language);
 
         Task SpecializeAsync();
 
-        bool ShutdownChannelIfExists(string language, string workerId);
+        Task<bool> ShutdownChannelIfExists(string language, string workerId);
```
Comment (yojagad) [inline src/WebJobs.Script/Rpc/IWebHostLanguageWorkerChannelManager.cs]: >ShutdownChannelIfExists [](start = 19, length = 23)

ShutdownChannelIfExistsAsync?

**Pull request merged**